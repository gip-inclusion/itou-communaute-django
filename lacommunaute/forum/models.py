from django.conf import settings
from django.contrib.contenttypes.fields import GenericRelation
from django.db import models
from django.urls import reverse
from django.utils.functional import cached_property
from machina.apps.forum.abstract_models import AbstractForum
from machina.models import DatedModel
from storages.backends.s3boto3 import S3Boto3Storage
from taggit.managers import TaggableManager

from lacommunaute.forum.enums import Kind as Forum_Kind
from lacommunaute.forum_conversation.models import Topic
from lacommunaute.forum_upvote.models import UpVote
from lacommunaute.partner.models import Partner
from lacommunaute.utils.validators import validate_image_size


class Forum(AbstractForum):
    kind = models.CharField(
        max_length=20, choices=Forum_Kind.choices, default=Forum_Kind.PUBLIC_FORUM, verbose_name="Type"
    )
    short_description = models.CharField(
        max_length=400, blank=True, null=True, verbose_name="Description courte (SEO)"
    )
    image = models.ImageField(
        storage=S3Boto3Storage(bucket_name=settings.AWS_STORAGE_BUCKET_NAME, file_overwrite=False),
        validators=[validate_image_size],
    )
    certified = models.BooleanField(default=False, verbose_name="Certifié par la communauté de l'inclusion")

    upvotes = GenericRelation(UpVote, related_query_name="forum")

    tags = TaggableManager()
    partner = models.ForeignKey(Partner, on_delete=models.CASCADE, null=True, blank=True)

    objects = models.Manager()

    def get_absolute_url(self):
        return reverse(
            "forum_extension:forum",
            kwargs={
                "pk": self.pk,
                "slug": self.slug,
            },
        )

    def get_unanswered_topics(self):
        return Topic.objects.unanswered().filter(forum__in=self.get_descendants(include_self=True))

    @cached_property
    def count_unanswered_topics(self):
        return self.get_unanswered_topics().count()

    def upvotes_count(self):
        return self.upvotes.count()

    @cached_property
    def is_in_documentation_area(self):
        return (self.type == Forum.FORUM_CAT and self.get_level() == 0) or (
            self.get_level() > 0 and self.get_ancestors().first().type == Forum.FORUM_CAT
        )

    @cached_property
    def is_toplevel_discussion_area(self):
        return self == Forum.objects.filter(lft=1, level=0).first()

    def get_session_rating(self, session_key):
        return getattr(ForumRating.objects.filter(forum=self, session_id=session_key).first(), "rating", None)

    def get_average_rating(self):
        return ForumRating.objects.filter(forum=self).aggregate(models.Avg("rating"))["rating__avg"]


class ForumRating(DatedModel):
    session_id = models.CharField(max_length=40)
    forum = models.ForeignKey(Forum, on_delete=models.CASCADE)
    rating = models.PositiveSmallIntegerField()
    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, null=True, blank=True)

    class Meta:
        verbose_name = "Notation Forum"
        verbose_name_plural = "Notations Forum"
        ordering = ("-created",)
