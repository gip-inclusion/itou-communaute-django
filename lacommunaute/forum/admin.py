from django.contrib import admin
from machina.apps.forum.admin import ForumAdmin as BaseForumAdmin

from lacommunaute.forum.models import ForumRating


class ForumAdmin(BaseForumAdmin):
    fieldsets = BaseForumAdmin.fieldsets
    fieldsets[0][1]["fields"] += ("short_description",)
    fieldsets[1][1]["fields"] += (
        "members_group",
        "invitation_token",
        "kind",
    )


@admin.register(ForumRating)
class ForumRatingAdmin(admin.ModelAdmin):
    list_display = ("forum", "rating", "created")
    list_filter = ("forum",)
    list_display_links = ("rating",)
    raw_id_fields = ("forum", "user")
