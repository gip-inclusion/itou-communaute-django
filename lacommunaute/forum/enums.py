from django.db import models


class Kind(models.TextChoices):
    PUBLIC_FORUM = "PUBLIC_FORUM", "Espace public"
    NEWS = "NEWS", "Actualit√©s"
