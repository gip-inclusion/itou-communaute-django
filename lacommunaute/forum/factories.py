import factory
from machina.core.db.models import get_model
from machina.test.factories.forum import ForumFactory as BaseForumFactory

from lacommunaute.forum.models import Forum
from lacommunaute.users.factories import GroupFactory


ForumPermission = get_model("forum_permission", "ForumPermission")
UserForumPermission = get_model("forum_permission", "UserForumPermission")


class ForumFactory(BaseForumFactory):
    type = Forum.FORUM_POST
    members_group = factory.SubFactory(GroupFactory, name=factory.SelfAttribute("..name"))
    name = factory.Faker("name")
    description = factory.Faker("sentence", nb_words=10)

    @factory.post_generation
    def with_public_perms(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        perms = [
            "can_see_forum",
            "can_read_forum",
            "can_start_new_topics",
            "can_reply_to_topics",
            "can_edit_own_posts",
            "can_delete_own_posts",
            "can_post_without_approval",
        ]

        anonymous_authorized_perms = [
            UserForumPermission(
                anonymous_user=True, authenticated_user=False, permission=permission, has_perm=True, forum=self
            )
            for permission in ForumPermission.objects.filter(codename__in=perms)
        ]

        authentified_authorized_perms = [
            UserForumPermission(
                anonymous_user=False, authenticated_user=True, permission=permission, has_perm=True, forum=self
            )
            for permission in ForumPermission.objects.filter(codename__in=perms)
        ]
        UserForumPermission.objects.bulk_create(anonymous_authorized_perms + authentified_authorized_perms)
