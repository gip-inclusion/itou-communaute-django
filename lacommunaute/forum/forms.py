import re

from django import forms
from django.conf import settings

from lacommunaute.forum.models import Forum


def wrap_iframe_in_div_tag(text):
    # iframe tags must be wrapped in a div tag to be displayed correctly
    # add div tag if not present

    iframe_regex = r"((<div>)?<iframe.*?</iframe>(</div>)?)"

    for match, starts_with, ends_with in re.findall(iframe_regex, text, re.DOTALL):
        if not starts_with and not ends_with:
            text = text.replace(match, f"<div>{match}</div>")

    return text


class ForumForm(forms.ModelForm):
    name = forms.CharField(required=True, label="Titre")
    short_description = forms.CharField(
        widget=forms.Textarea(attrs={"rows": 3}),
        max_length=400,
        required=True,
        label="Sous-titre (400 caractères pour le SEO)",
    )
    description = forms.CharField(
        widget=forms.Textarea(attrs={"rows": 20}), required=False, label="Contenu (markdown autorisé)"
    )
    image = forms.ImageField(
        required=False,
        label="Banniere de couverture, format 1200 x 630 pixels recommandé",
        widget=forms.FileInput(attrs={"accept": settings.SUPPORTED_IMAGE_FILE_TYPES.keys()}),
    )

    def save(self, commit=True):
        forum = super().save(commit=False)
        forum.description = wrap_iframe_in_div_tag(self.cleaned_data.get("description"))
        if commit:
            forum.save()
        return forum

    class Meta:
        model = Forum
        fields = ["name", "short_description", "description", "image"]
