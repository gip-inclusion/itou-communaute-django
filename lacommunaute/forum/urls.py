from django.urls import path
from machina.apps.forum.views import IndexView

from lacommunaute.forum.views import CategoryForumListView, ForumView


app_name = "forum_extension"


urlpatterns = [
    path("forum/<str:slug>-<int:pk>/", ForumView.as_view(), name="forum"),
    path("forum/documentation/", CategoryForumListView.as_view(), name="documentation"),
    path("forum/", IndexView.as_view(), name="index"),
]
