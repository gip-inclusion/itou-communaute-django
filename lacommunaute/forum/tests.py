from dateutil.relativedelta import relativedelta
from django.contrib.auth.models import AnonymousUser
from django.db import IntegrityError
from django.test import RequestFactory, TestCase
from django.urls import reverse
from django.utils import timezone
from django.utils.http import urlencode
from faker import Faker
from machina.core.db.models import get_model
from machina.core.loading import get_class

from lacommunaute.forum.factories import ForumFactory
from lacommunaute.forum.views import ForumView
from lacommunaute.forum_conversation.factories import PostFactory, TopicFactory
from lacommunaute.forum_conversation.forms import PostForm
from lacommunaute.forum_conversation.forum_polls.factories import TopicPollVoteFactory
from lacommunaute.forum_conversation.models import Topic
from lacommunaute.forum_upvote.factories import UpVoteFactory
from lacommunaute.users.factories import UserFactory


faker = Faker()

ForumPermission = get_model("forum_permission", "ForumPermission")
UserForumPermission = get_model("forum_permission", "UserForumPermission")
PermissionHandler = get_class("forum_permission.handler", "PermissionHandler")
assign_perm = get_class("forum_permission.shortcuts", "assign_perm")
remove_perm = get_class("forum_permission.shortcuts", "remove_perm")


class ForumViewQuerysetTest(TestCase):
    @classmethod
    def setUpTestData(cls):
        cls.user = UserFactory()
        cls.forum = ForumFactory()
        cls.view = ForumView()
        cls.view.kwargs = {"pk": cls.forum.pk}
        cls.view.request = RequestFactory().get("/")
        cls.view.request.user = cls.user

    def test_excluded_announces_topics(self):
        TopicFactory(forum=self.forum, poster=self.user, type=Topic.TOPIC_ANNOUNCE)
        self.assertFalse(self.view.get_queryset())

    def test_exclude_not_approved_posts(self):
        TopicFactory(forum=self.forum, poster=self.user, approved=False)
        self.assertFalse(self.view.get_queryset())

    def test_ordering_topics_on_last_post(self):
        old_topic = TopicFactory(forum=self.forum, poster=self.user)
        new_topic = TopicFactory(forum=self.forum, poster=self.user)

        PostFactory(topic=old_topic, poster=self.user)
        PostFactory(topic=new_topic, poster=self.user)

        qs = self.view.get_queryset()
        self.assertEqual(qs.first(), new_topic)
        self.assertEqual(qs.last(), old_topic)

        PostFactory(topic=old_topic, poster=self.user)
        self.assertEqual(qs.first(), old_topic)
        self.assertEqual(qs.last(), new_topic)

    def test_pagination(self):
        self.assertEqual(10, self.view.paginate_by)

    def test_has_liked(self):
        TopicFactory(forum=self.forum, poster=self.user, with_like=True)

        first_topic = self.view.get_queryset().first()
        self.assertEqual(first_topic.likes, 1)
        self.assertTrue(first_topic.has_liked)

    def test_has_not_liked(self):
        TopicFactory(forum=self.forum, poster=self.user)

        first_topic = self.view.get_queryset().first()
        self.assertEqual(first_topic.likes, 0)
        self.assertFalse(first_topic.has_liked)


class ForumViewTest(TestCase):
    @classmethod
    def setUpTestData(cls):
        cls.topic = TopicFactory(with_post=True)
        cls.user = cls.topic.poster
        cls.forum = cls.topic.forum

        # Assign some permissions
        cls.perm_handler = PermissionHandler()
        assign_perm("can_read_forum", cls.user, cls.forum)
        assign_perm("can_see_forum", cls.user, cls.forum)
        assign_perm("can_post_without_approval", cls.user, cls.forum)
        assign_perm("can_reply_to_topics", cls.user, cls.forum)

        cls.url = reverse("forum:forum", kwargs={"pk": cls.forum.pk, "slug": cls.forum.slug})

    def test_subscription_button_is_hidden(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertContains(
            response,
            reverse(
                "forum_conversation_extension:post_create",
                kwargs={
                    "forum_pk": self.forum.pk,
                    "forum_slug": self.forum.slug,
                    "pk": self.topic.pk,
                    "slug": self.topic.slug,
                },
            ),
        )
        self.assertNotContains(
            response,
            reverse(
                "forum_member:topic_subscribe",
                kwargs={
                    "pk": self.topic.pk,
                },
            ),
        )

    def test_show_comments(self):
        topic_url = reverse(
            "forum_conversation_extension:showmore_posts",
            kwargs={
                "forum_pk": self.forum.pk,
                "forum_slug": self.forum.slug,
                "pk": self.topic.pk,
                "slug": self.topic.slug,
            },
        )
        self.client.force_login(self.user)
        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertNotContains(response, f'<a href="{topic_url}"')

        PostFactory.create_batch(2, topic=self.topic, poster=self.user)
        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertContains(response, f'hx-get="{topic_url}"')
        self.assertContains(response, "Voir les 2 autres rÃ©ponses")

    def test_show_more_content(self):
        topic = TopicFactory(
            with_post=True, poster=self.user, forum=self.forum, post__content=faker.paragraph(nb_sentences=100)
        )
        topic_url = reverse(
            "forum_conversation_extension:showmore_topic",
            kwargs={
                "forum_pk": topic.forum.pk,
                "forum_slug": topic.forum.slug,
                "pk": topic.pk,
                "slug": topic.slug,
            },
        )
        self.client.force_login(self.user)
        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertContains(response, f'hx-get="{topic_url}"')
        self.assertContains(response, "+ voir la suite")

    def test_has_liked(self):
        TopicFactory(forum=self.forum, poster=self.user, with_post=True, with_like=True)

        self.client.force_login(self.user)
        response = self.client.get(self.url)
        # icon: solid heart
        self.assertContains(response, '<i class="ri-heart-3-fill" aria-hidden="true"></i>')
        self.assertContains(response, "<span>1 J'aime</span>")

    def test_has_not_liked(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        # icon: regular heart (outlined)
        self.assertContains(response, '<i class="ri-heart-3-line" aria-hidden="true"></i>')
        self.assertContains(response, "<span>0 J'aime</span>")

    def test_anonymous_like(self):
        assign_perm("can_read_forum", AnonymousUser(), self.topic.forum)
        params = {"next_url": self.url}
        url = f"{reverse('inclusion_connect:authorize')}?{urlencode(params)}"

        response = self.client.get(self.url)
        self.assertContains(response, url)

        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertNotContains(response, url)

    def test_moderator_links(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertNotContains(
            response,
            reverse(
                "forum_extension:engagement",
                kwargs={"pk": self.forum.pk, "slug": self.forum.slug},
            ),
        )
        self.assertNotContains(
            response,
            reverse(
                "members:forum_profiles",
                kwargs={"pk": self.forum.pk, "slug": self.forum.slug},
            ),
        )

        assign_perm("can_approve_posts", self.user, self.forum)
        response = self.client.get(self.url)
        self.assertContains(
            response,
            reverse(
                "forum_extension:engagement",
                kwargs={"pk": self.forum.pk, "slug": self.forum.slug},
            ),
        )
        self.assertContains(
            response,
            reverse(
                "members:forum_profiles",
                kwargs={"pk": self.forum.pk, "slug": self.forum.slug},
            ),
        )

    def test_poll_form(self):
        topic = TopicFactory(forum=self.forum, poster=self.user, with_post=True, with_poll_vote=True)
        poll_option = topic.poll.options.first()
        self.client.force_login(self.user)

        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 200)
        self.assertContains(response, poll_option.poll.question)
        self.assertContains(response, poll_option.text)

    def test_postform_in_context(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 200)
        self.assertIsInstance(response.context_data["form"], PostForm)
        self.assertContains(response, f'id="collapsePost{self.topic.pk}')

    def test_cannot_submit_post(self):
        user = UserFactory()
        assign_perm("can_read_forum", user, self.forum)
        assign_perm("can_see_forum", user, self.forum)
        remove_perm("can_reply_to_topics", user, self.forum)
        self.client.force_login(user)

        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertNotContains(response, f'id="collapsePost{self.topic.pk}')


class ForumModelTest(TestCase):
    def test_invitation_token_is_unique(self):
        forum = ForumFactory()

        with self.assertRaises(IntegrityError):
            forum.id = None
            forum.save()

    def test_get_stats(self):

        forum = ForumFactory()
        topic = TopicFactory(forum=forum, with_post=True, with_poll_vote=True)
        UpVoteFactory(post=topic.posts.first(), voter=topic.poster)

        # create old instance to ensure they are filtered
        fifteen_days_ago = timezone.now() - relativedelta(days=15)
        old_topic = TopicFactory(forum=forum)
        old_topic.created = fifteen_days_ago
        old_topic.save()
        old_post = PostFactory(topic=old_topic, poster=old_topic.poster)
        old_post.created = fifteen_days_ago
        old_post.save()
        old_pollvote = TopicPollVoteFactory(voter=old_topic.poster, poll_option__poll__topic=old_topic)
        old_pollvote.timestamp = fifteen_days_ago
        old_pollvote.save()
        old_upvote = UpVoteFactory(post=old_topic.posts.first(), voter=old_topic.poster)
        old_upvote.created_at = fifteen_days_ago
        old_upvote.save()

        # create fake forum to ensure post, topic, etcâ¦ number is filter by forum
        fake_forum = ForumFactory()
        fake_topic = TopicFactory(forum=fake_forum, with_post=True, with_poll_vote=True)
        UpVoteFactory(post=fake_topic.posts.first(), voter=fake_topic.poster)

        self.assertEqual(
            forum.get_stats(1),
            {
                "period": [timezone.now().strftime("%Y-%W")],
                "pollvotes": [1],
                "upvotes": [1],
                "posts": [1],
                "topics": [1],
            },
        )

    def test_count_engaged_users(self):
        # first user posts, likes, votes
        topic = TopicFactory(with_post=True, with_like=True, with_poll_vote=True)

        # second user posts
        PostFactory(topic=topic, with_upvote=True)

        # anonymous user posts
        PostFactory(topic=topic, anonymous=True)

        # post, like, vote in an other forum = ignored in count
        TopicFactory(with_post=True, with_like=True, with_poll_vote=True)

        self.assertEqual(
            topic.forum.count_engaged_users,
            {
                "posters": 2,
                "likers": 1,
                "voters": 1,
                "upvoters": 1,
                "authenticated_users": 2,
                "anonymous_posters": 1,
                "all_users": 3,
            },
        )
