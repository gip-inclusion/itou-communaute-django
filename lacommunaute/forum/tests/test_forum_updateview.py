from io import BytesIO

import pytest
from django.conf import settings
from django.core.files.uploadedfile import SimpleUploadedFile
from django.urls import reverse
from faker import Faker
from PIL import Image
from pytest_django.asserts import assertContains
from taggit.models import Tag

from lacommunaute.forum.factories import CategoryForumFactory, ForumFactory
from lacommunaute.forum.forms import SubCategoryForumUpdateForm
from lacommunaute.partner.factories import PartnerFactory
from lacommunaute.users.factories import UserFactory
from lacommunaute.utils.testing import parse_response_to_soup, reset_model_sequence_fixture


faker = Faker(settings.LANGUAGE_CODE)

reset_tag_sequence = pytest.fixture(reset_model_sequence_fixture(Tag))


@pytest.fixture
def fake_image():
    fake = Faker()
    image_name = fake.pystr(min_chars=30, max_chars=40, prefix="pytest_", suffix=".jpg")

    image = Image.new("RGB", (100, 100))
    image_file = BytesIO()
    image.save(image_file, format="JPEG")
    image_file.seek(0)
    uploaded_image = SimpleUploadedFile(image_name, image_file.read(), content_type="image/jpeg")

    yield uploaded_image


def test_user_access(client, db):
    forum = ForumFactory()
    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})
    response = client.get(url)
    assert response.status_code == 302

    user = UserFactory()
    client.force_login(user)
    response = client.get(url)
    assert response.status_code == 403

    user.is_staff = True
    user.save()
    response = client.get(url)
    assert response.status_code == 200


@pytest.mark.parametrize(
    "forum,should_be_subcategoryupdateform",
    [
        (lambda: ForumFactory(), False),
        (lambda: CategoryForumFactory(), False),
        (lambda: CategoryForumFactory(with_child=True).get_children().first(), True),
    ],
)
def test_context_data(client, db, forum, should_be_subcategoryupdateform):
    client.force_login(UserFactory(is_in_staff_group=True))
    forum = forum()
    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})
    response = client.get(url)
    assertContains(response, f"Mettre Ã  jour le forum {forum.name}", html=True)
    assertContains(response, reverse("forum_extension:forum", kwargs={"pk": forum.pk, "slug": forum.slug}))
    assert isinstance(response.context["form"], SubCategoryForumUpdateForm) == should_be_subcategoryupdateform


def test_update_forum_image(client, db, fake_image):
    client.force_login(UserFactory(is_in_staff_group=True))
    forum = CategoryForumFactory(with_child=True).get_children().first()
    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})

    response = client.post(
        url,
        data={
            "name": "new name",
            "short_description": "new short description",
            "description": "new description",
            "image": fake_image,
        },
    )
    assert response.status_code == 302

    forum.refresh_from_db()
    assert forum.name == "new name"
    assert forum.short_description == "new short description"
    assert forum.description.raw == "new description"
    assert forum.image.name == fake_image.name


def test_update_subcategory_forum_parent(client, db):
    forum = CategoryForumFactory(with_child=True).get_children().first()
    new_parent = CategoryForumFactory()
    client.force_login(UserFactory(is_in_staff_group=True))

    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})
    response = client.post(
        url,
        data={
            "name": forum.name,
            "short_description": forum.short_description,
            "description": forum.description.raw,
            "parent": new_parent.pk,
        },
    )
    assert response.status_code == 302

    forum.refresh_from_db()
    assert forum.parent == new_parent


def test_certified_forum(client, db):
    client.force_login(UserFactory(is_in_staff_group=True))
    forum = CategoryForumFactory(with_child=True).get_children().first()
    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})

    response = client.post(
        url,
        data={
            "name": "new name",
            "short_description": "new short description",
            "description": "new description",
            "certified": True,
        },
    )
    assert response.status_code == 302

    forum.refresh_from_db()
    assert forum.certified is True


def test_selected_tags_are_preloaded(client, db, reset_tag_sequence, snapshot):
    client.force_login(UserFactory(is_in_staff_group=True))
    forum = ForumFactory(with_tags=["iae", "siae", "prescripteur"])
    Tag.objects.create(name="undesired_tag")
    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})

    response = client.get(url)
    assert response.status_code == 200

    content_tags = parse_response_to_soup(
        response, selector="#div_id_tags", replace_in_href=[tag for tag in Tag.objects.all()]
    )
    assert str(content_tags) == snapshot(name="selected_tags_preloaded")


def test_added_tags_are_saved(client, db):
    client.force_login(UserFactory(is_in_staff_group=True))
    forum = ForumFactory()

    Tag.objects.bulk_create([Tag(name=tag, slug=tag) for tag in [faker.word() for _ in range(3)]])
    tags_list = [faker.word() for i in range(2)]

    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})
    response = client.post(
        url,
        data={
            "name": forum.name,
            "short_description": forum.short_description,
            "description": forum.description.raw,
            "tags": [Tag.objects.first().pk],
            "new_tags": ", ".join(tags_list),
        },
    )

    assert response.status_code == 302

    forum.refresh_from_db()
    assert all(tag in [tag.name for tag in forum.tags.all()] for tag in [Tag.objects.first().name] + tags_list)


def test_update_forum_without_tag(client, db):
    client.force_login(UserFactory(is_in_staff_group=True))
    forum = ForumFactory()
    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})
    response = client.post(
        url,
        data={
            "name": forum.name,
            "short_description": forum.short_description,
            "description": forum.description.raw,
        },
    )
    assert response.status_code == 302

    forum.refresh_from_db()
    assert forum.tags.count() == 0


@pytest.mark.parametrize(
    "initial_partner,post_partner",
    [
        (lambda: None, lambda: None),
        (lambda: None, lambda: PartnerFactory()),
        (lambda: PartnerFactory(), lambda: None),
        (lambda: PartnerFactory(), lambda: PartnerFactory()),
    ],
)
def test_select_partner(client, db, initial_partner, post_partner):
    initial_partner = initial_partner()
    post_partner = post_partner()
    forum = ForumFactory(partner=initial_partner)
    client.force_login(UserFactory(is_in_staff_group=True))

    url = reverse("forum_extension:edit_forum", kwargs={"pk": forum.pk, "slug": forum.slug})
    response = client.post(
        url,
        data={
            "name": forum.name,
            "short_description": forum.short_description,
            "description": forum.description.raw,
            "partner": post_partner.id if post_partner else "",
        },
    )
    assert response.status_code == 302

    forum.refresh_from_db()
    assert forum.partner == post_partner
