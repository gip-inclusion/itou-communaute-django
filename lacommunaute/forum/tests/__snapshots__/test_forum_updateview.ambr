# serializer version: 1
# name: test_selected_tags_are_preloaded[selected_tags_preloaded]
  '''
  <div class="form-group" id="div_id_tags">
      
          
              <label class="control-label" for="id_tags">
                  SÃ©lectionner un ou plusieurs tags
                  
                      <span class="text-muted">(optional)</span>
                  
              </label>
          
          
              
                  <div class="form-check form-check-inline">
                      <input checked="" class="form-check-input" id="id_tags_0" name="tags" type="checkbox" value="9999"/>
                      <label class="form-check-label tag bg-info-lighter text-info" for="id_tags_0">
                          iae
                      </label>
                  </div>
              
                  <div class="form-check form-check-inline">
                      <input checked="" class="form-check-input" id="id_tags_1" name="tags" type="checkbox" value="10000"/>
                      <label class="form-check-label tag bg-info-lighter text-info" for="id_tags_1">
                          siae
                      </label>
                  </div>
              
                  <div class="form-check form-check-inline">
                      <input checked="" class="form-check-input" id="id_tags_2" name="tags" type="checkbox" value="10001"/>
                      <label class="form-check-label tag bg-info-lighter text-info" for="id_tags_2">
                          prescripteur
                      </label>
                  </div>
              
                  <div class="form-check form-check-inline">
                      <input class="form-check-input" id="id_tags_3" name="tags" type="checkbox" value="10002"/>
                      <label class="form-check-label tag bg-info-lighter text-info" for="id_tags_3">
                          undesired_tag
                      </label>
                  </div>
              
          
      
      
      
      
  </div>
  '''
# ---
